modin/core/storage_formats/pandas/query_compiler_caster.py

```diff
@@ -0,0 +1,159 @@
+        apply_argument_cast(cls)
+
+
+def cast_nested_args_to_current_qc_type(arguments, current_qc):
+    """
+    Cast all arguments in nested fashion to current query compiler.
+
+    Parameters
+    ----------
+    arguments : tuple or dict
+    current_qc : BaseQueryCompiler
+
+    Returns
+    -------
+    tuple or dict
+        Returns args and kwargs with all query compilers casted to current_qc.
+    """
+
+    def cast_arg_to_current_qc(arg):
+        current_qc_type = type(current_qc)
+        if isinstance(arg, BaseQueryCompiler) and not isinstance(arg, current_qc_type):
+            data_cls = current_qc._modin_frame
+            return current_qc_type.from_pandas(arg.to_pandas(), data_cls)
+        else:
+            return arg
+
+    imutable_types = (FrozenList, tuple)
+    if isinstance(arguments, imutable_types):
+        args_type = type(arguments)
+        arguments = list(arguments)
+        arguments = cast_nested_args_to_current_qc_type(arguments, current_qc)
+
+        return args_type(arguments)
+    if isinstance(arguments, list):
+        for i in range(len(arguments)):
+            if isinstance(arguments[i], (list, dict)):
+                cast_nested_args_to_current_qc_type(arguments[i], current_qc)
+            else:
+                arguments[i] = cast_arg_to_current_qc(arguments[i])
+    elif isinstance(arguments, dict):
+        for key in arguments:
+            if isinstance(arguments[key], (list, dict)):
+                cast_nested_args_to_current_qc_type(arguments[key], current_qc)
+            else:
+                arguments[key] = cast_arg_to_current_qc(arguments[key])
+    return arguments
+
+
+def apply_argument_cast(obj: Fn) -> Fn:
+    """
+    Cast all arguments that are query compilers to the current query compiler.
+
+    Parameters
+    ----------
+    obj : function
+
+    Returns
+    -------
+    function
+        Returns decorated function which does argument casting.
+    """
+    if isinstance(obj, type):
+        all_attrs = dict(inspect.getmembers(obj))
+        all_attrs.pop("__abstractmethods__")
+
+        # This is required because inspect converts class methods to member functions
+        current_class_attrs = vars(obj)
+        for key in current_class_attrs:
+            all_attrs[key] = current_class_attrs[key]
+
+        for attr_name, attr_value in all_attrs.items():
+            if isinstance(
+                attr_value, (FunctionType, MethodType, classmethod, staticmethod)
+            ):
+                wrapped = apply_argument_cast(attr_value)
+                setattr(obj, attr_name, wrapped)
+        return obj  # type: ignore [return-value]
+    elif isinstance(obj, classmethod):
+        return classmethod(apply_argument_cast(obj.__func__))  # type: ignore [return-value, arg-type]
+    elif isinstance(obj, staticmethod):
+        return staticmethod(apply_argument_cast(obj.__func__))
+
+    @functools.wraps(obj)
+    def cast_args(*args: Tuple, **kwargs: Dict) -> Any:
```

Comment:
@arunjose696 doesn't this function break type hints?

Reply:
I am not sure how this would break type hints, as this function is similar to run_and_log function used in modin logging which would also be wrapping the qc layer.

Reply:
> I am not clear how this would break type hints, as this function is similar to run_and_log function used in modin logging which would also be wrapping the qc layer

It should be visible in IDE. Do you see problems with it or not?

Reply:
![image](https://github.com/user-attachments/assets/f1d192df-e92c-4db0-8b3a-0f19e9801b18)
I can see it in IDE

Reply:
Thanks for checking. `| Any` seems redundant, someone will need to sort this out in the future.